File: common/src/main/java/link/e4steam/MinecraftUiCompat.java
package link.e4steam;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.components.Button;
import net.minecraft.client.gui.screens.ConnectScreen;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.client.multiplayer.ServerData;
import net.minecraft.client.multiplayer.resolver.ServerAddress;
import net.minecraft.network.chat.Component;

import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.lang.reflect.Proxy;
import java.util.Arrays;
import java.util.Comparator;
import java.util.Optional;
import java.util.function.Consumer;

/**
 * Small reflection boundary for GUI APIs that changed after Minecraft 1.17.
 *
 * <p>The main JAR is compiled against 1.20.2, just like upstream e4steam's wide
 * compatibility build. Keeping Button.Builder and Tooltip out of bytecode
 * descriptors lets the same classes load on older Minecraft releases.</p>
 */
public final class MinecraftUiCompat {
    private static final String[] MESSAGE_SCREEN_CLASS_NAMES = {
            // Mojang mappings through 1.20.6.
            "net.minecraft.client.gui.screens.GenericDirtMessageScreen",
            // Mojang mappings from 1.21 onward.
            "net.minecraft.client.gui.screens.GenericMessageScreen",
            // Stable Fabric intermediary name used across both eras.
            "net.minecraft.class_424"
    };

    private static final String[] DISCONNECT_METHOD_NAMES = {
            "disconnect",
            "clearLevel",
            "method_18099",
            "method_18096",
            "m_91387_",
            "m_91398_"
    };

    private MinecraftUiCompat() {
    }

    public static Button button(
            Component message,
            Consumer<Button> onPress,
            int x,
            int y,
            int width,
            int height
    ) {
        Button fromBuilder = tryBuilder(message, onPress, x, y, width, height);
        if (fromBuilder != null) {
            return fromBuilder;
        }

        Button fromConstructor = tryConstructor(message, onPress, x, y, width, height);
        if (fromConstructor != null) {
            return fromConstructor;
        }

        throw new IllegalStateException("No compatible Minecraft Button factory was found");
    }

    /**
     * Adds a modern hover tooltip where that API exists. Minecraft 1.17-1.19.2
     * simply keeps the button without a tooltip.
     */
    public static void tooltip(Object widget, Component message) {
        if (widget == null || message == null) {
            return;
        }

        for (Method setter : widget.getClass().getMethods()) {
            if (setter.getParameterCount() != 1 || Modifier.isStatic(setter.getModifiers())) {
                continue;
            }
            Class<?> tooltipType = setter.getParameterTypes()[0];
            Object tooltip = createTooltip(tooltipType, message);
            if (tooltip == null) {
                continue;
            }
            try {
                setter.invoke(widget, tooltip);
            } catch (ReflectiveOperationException ignored) {
                // A tooltip is optional; the button itself remains usable.
                continue;
            }
            return;
        }
    }

    /**
     * Creates Minecraft's generic progress/message screen without linking the
     * caller to the Mojang class name that changed in 1.21.
     */
    public static Screen messageScreen(Component message, Screen fallback) {
        for (String className : MESSAGE_SCREEN_CLASS_NAMES) {
            try {
                Class<?> screenType = Class.forName(className, false, Screen.class.getClassLoader());
                if (!Screen.class.isAssignableFrom(screenType)) {
                    continue;
                }
                for (Constructor<?> constructor : screenType.getDeclaredConstructors()) {
                    Class<?>[] parameterTypes = constructor.getParameterTypes();
                    if (parameterTypes.length != 1
                            || !parameterTypes[0].isAssignableFrom(message.getClass())) {
                        continue;
                    }
                    constructor.setAccessible(true);
                    return (Screen) constructor.newInstance(message);
                }
            } catch (ClassNotFoundException ignored) {
                // Try the next mapping/runtime name.
            } catch (ReflectiveOperationException | RuntimeException failure) {
                // A progress screen is cosmetic. Keep joining via the previous
                // screen if this Minecraft version changes its constructor.
                break;
            }
        }
        return fallback;
    }

    /**
     * Reads the active screen across the Minecraft 26.x move from
     * {@code Minecraft.screen} to {@code Minecraft.gui.screen()}.
     */
    public static Screen currentScreen(Minecraft minecraft) {
        try {
            return minecraft.screen;
        } catch (NoSuchFieldError ignored) {
            try {
                Method getter = minecraft.gui.getClass().getMethod("screen");
                return (Screen) getter.invoke(minecraft.gui);
            } catch (ReflectiveOperationException failure) {
                throw new IllegalStateException("No compatible current-screen accessor was found", failure);
            }
        }
    }

    /** Opens a screen across the Minecraft 26.x setScreenAndShow rename. */
    public static void setScreen(Minecraft minecraft, Screen screen) {
        try {
            minecraft.setScreen(screen);
        } catch (NoSuchMethodError ignored) {
            try {
                Method setter = minecraft.getClass().getMethod("setScreenAndShow", Screen.class);
                setter.invoke(minecraft, screen);
            } catch (ReflectiveOperationException failure) {
                throw new IllegalStateException("No compatible screen setter was found", failure);
            }
        }
    }

    /** Adds a chat message after Minecraft 26.x moved chat from Gui to Hud. */
    public static void addChatMessage(Minecraft minecraft, Component message) {
        try {
            minecraft.gui.getChat().addMessage(message);
        } catch (NoSuchMethodError ignored) {
            try {
                Object hud = minecraft.gui.getClass().getField("hud").get(minecraft.gui);
                Object chat = hud.getClass().getMethod("getChat").invoke(hud);
                Method addMessage = Arrays.stream(chat.getClass().getMethods())
                        .filter(method -> method.getName().equals("addClientSystemMessage")
                                || method.getName().equals("addMessage"))
                        .filter(method -> method.getParameterCount() == 1)
                        .filter(method -> method.getParameterTypes()[0].isAssignableFrom(message.getClass()))
                        .findFirst()
                        .orElseThrow(() -> new NoSuchMethodException("ChatComponent system-message method"));
                addMessage.invoke(chat, message);
            } catch (ReflectiveOperationException failure) {
                throw new IllegalStateException("No compatible chat message API was found", failure);
            }
        }
    }

    /**
     * Starts Minecraft's normal connection screen across its old four-argument
     * and newer five-/six-argument signatures.
     */
    public static void connect(
            Screen parent,
            Minecraft minecraft,
            ServerAddress address,
            String displayName,
            String endpoint
    ) throws ReflectiveOperationException {
        Object serverData = createServerData(displayName, endpoint);
        Method[] methods = ConnectScreen.class.getDeclaredMethods();
        Arrays.sort(methods, Comparator
                .<Method>comparingInt(method -> Modifier.isPublic(method.getModifiers()) ? 0 : 1)
                .thenComparingInt(Method::getParameterCount));
        for (Method method : methods) {
            if (!Modifier.isStatic(method.getModifiers()) || method.getReturnType() != void.class) {
                continue;
            }
            Class<?>[] types = method.getParameterTypes();
            if (!containsAssignable(types, Screen.class)
                    || !containsAssignable(types, Minecraft.class)
                    || !containsAssignable(types, ServerAddress.class)
                    || !containsAssignable(types, ServerData.class)) {
                continue;
            }

            Object[] arguments = new Object[types.length];
            boolean valid = true;
            for (int i = 0; i < types.length; i++) {
                Class<?> type = types[i];
                if (type.isInstance(parent) || type == Screen.class) {
                    arguments[i] = parent;
                } else if (type.isInstance(minecraft) || type == Minecraft.class) {
                    arguments[i] = minecraft;
                } else if (type.isInstance(address) || type == ServerAddress.class) {
                    arguments[i] = address;
                } else if (type.isInstance(serverData) || type == ServerData.class) {
                    arguments[i] = serverData;
                } else if (type == boolean.class) {
                    arguments[i] = false;
                } else if (type == int.class) {
                    arguments[i] = 0;
                } else if (type == long.class) {
                    arguments[i] = 0L;
                } else if (type == Optional.class) {
                    arguments[i] = Optional.empty();
                } else if (!type.isPrimitive()) {
                    arguments[i] = null;
                } else {
                    valid = false;
                    break;
                }
            }
            if (!valid) {
                continue;
            }
            method.setAccessible(true);
            method.invoke(null, arguments);
            return;
        }
        throw new NoSuchMethodException("No compatible ConnectScreen entry point was found");
    }

    /** Disconnects the current world across the old clearLevel/new disconnect rename. */
    public static void disconnect(Minecraft minecraft, Screen nextScreen)
            throws ReflectiveOperationException {
        for (String name : DISCONNECT_METHOD_NAMES) {
            for (Method method : Minecraft.class.getMethods()) {
                if (!method.getName().equals(name)
                        || Modifier.isStatic(method.getModifiers())
                        || method.getReturnType() != void.class
                        || method.getParameterCount() != 1
                        || !method.getParameterTypes()[0].isAssignableFrom(nextScreen.getClass())) {
                    continue;
                }
                method.invoke(minecraft, nextScreen);
                return;
            }
        }
        throw new NoSuchMethodException("No compatible Minecraft disconnect method was found");
    }

    private static Button tryBuilder(
            Component message,
            Consumer<Button> onPress,
            int x,
            int y,
            int width,
            int height
    ) {
        for (Method factory : Button.class.getDeclaredMethods()) {
            if (!Modifier.isStatic(factory.getModifiers()) || factory.getParameterCount() != 2) {
                continue;
            }
            Class<?>[] parameters = factory.getParameterTypes();
            if (!parameters[0].isAssignableFrom(message.getClass())
                    && !parameters[0].isAssignableFrom(Component.class)) {
                continue;
            }
            if (!parameters[1].isInterface()) {
                continue;
            }

            Object callback = callback(parameters[1], onPress, message);
            try {
                factory.setAccessible(true);
                Object builder = factory.invoke(null, message, callback);
                if (builder == null || builder instanceof Button) {
                    continue;
                }

                Method bounds = Arrays.stream(builder.getClass().getMethods())
                        .filter(method -> !Modifier.isStatic(method.getModifiers()))
                        .filter(method -> method.getParameterCount() == 4)
                        .filter(method -> Arrays.stream(method.getParameterTypes())
                                .allMatch(type -> type == int.class))
                        .filter(method -> method.getReturnType().isAssignableFrom(builder.getClass())
                                || builder.getClass().isAssignableFrom(method.getReturnType()))
                        .findFirst()
                        .orElse(null);
                if (bounds == null) {
                    continue;
                }
                bounds.invoke(builder, x, y, width, height);

                Method build = Arrays.stream(builder.getClass().getMethods())
                        .filter(method -> !Modifier.isStatic(method.getModifiers()))
                        .filter(method -> method.getParameterCount() == 0)
                        .filter(method -> Button.class.isAssignableFrom(method.getReturnType()))
                        .findFirst()
                        .orElse(null);
                if (build != null) {
                    return (Button) build.invoke(builder);
                }
            } catch (ReflectiveOperationException | RuntimeException ignored) {
                // Fall through to the pre-builder constructor path.
            }
        }
        return null;
    }

    private static Object createServerData(String displayName, String endpoint)
            throws ReflectiveOperationException {
        Constructor<?>[] constructors = ServerData.class.getDeclaredConstructors();
        Arrays.sort(constructors, Comparator.comparingInt(Constructor::getParameterCount));
        ReflectiveOperationException lastFailure = null;
        for (Constructor<?> constructor : constructors) {
            Class<?>[] types = constructor.getParameterTypes();
            if (types.length < 2 || types[0] != String.class || types[1] != String.class) {
                continue;
            }
            Object[] arguments = new Object[types.length];
            arguments[0] = displayName;
            arguments[1] = endpoint;
            for (int i = 2; i < types.length; i++) {
                Class<?> type = types[i];
                if (type == boolean.class) {
                    arguments[i] = false;
                } else if (type == int.class) {
                    arguments[i] = 0;
                } else if (type == long.class) {
                    arguments[i] = 0L;
                } else if (type.isEnum()) {
                    Object[] constants = type.getEnumConstants();
                    Object selected = null;
                    for (Object constant : constants) {
                        if (((Enum<?>) constant).name().equals("OTHER")) {
                            selected = constant;
                            break;
                        }
                    }
                    arguments[i] = selected != null
                            ? selected
                            : (constants.length == 0 ? null : constants[0]);
                } else if (type == Optional.class) {
                    arguments[i] = Optional.empty();
                } else {
                    arguments[i] = null;
                }
            }
            try {
                constructor.setAccessible(true);
                return constructor.newInstance(arguments);
            } catch (ReflectiveOperationException failure) {
                lastFailure = failure;
            }
        }
        if (lastFailure != null) {
            throw lastFailure;
        }
        throw new NoSuchMethodException("No compatible ServerData constructor was found");
    }

    private static boolean containsAssignable(Class<?>[] haystack, Class<?> needle) {
        for (Class<?> type : haystack) {
            if (type != Object.class
                    && (type.isAssignableFrom(needle) || needle.isAssignableFrom(type))) {
                return true;
            }
        }
        return false;
    }

    private static Button tryConstructor(
            Component message,
            Consumer<Button> onPress,
            int x,
            int y,
            int width,
            int height
    ) {
        Constructor<?>[] constructors = Button.class.getDeclaredConstructors();
        Arrays.sort(constructors, Comparator.comparingInt(Constructor::getParameterCount));
        for (Constructor<?> constructor : constructors) {
            Class<?>[] parameterTypes = constructor.getParameterTypes();
            if (parameterTypes.length < 6
                    || parameterTypes[0] != int.class
                    || parameterTypes[1] != int.class
                    || parameterTypes[2] != int.class
                    || parameterTypes[3] != int.class
                    || !Component.class.isAssignableFrom(parameterTypes[4])
                    || !parameterTypes[5].isInterface()) {
                continue;
            }

            Object[] arguments = new Object[parameterTypes.length];
            arguments[0] = x;
            arguments[1] = y;
            arguments[2] = width;
            arguments[3] = height;
            arguments[4] = message;
            arguments[5] = callback(parameterTypes[5], onPress, message);
            for (int i = 6; i < parameterTypes.length; i++) {
                arguments[i] = defaultValue(parameterTypes[i], message);
            }

            try {
                constructor.setAccessible(true);
                return (Button) constructor.newInstance(arguments);
            } catch (ReflectiveOperationException | RuntimeException ignored) {
                // Try the next constructor shape.
            }
        }
        return null;
    }

    private static Object createTooltip(Class<?> tooltipType, Component message) {
        for (Method method : tooltipType.getDeclaredMethods()) {
            if (!Modifier.isStatic(method.getModifiers())
                    || method.getParameterCount() != 1
                    || !tooltipType.isAssignableFrom(method.getReturnType())
                    || !method.getParameterTypes()[0].isAssignableFrom(message.getClass())) {
                continue;
            }
            try {
                method.setAccessible(true);
                return method.invoke(null, message);
            } catch (ReflectiveOperationException | RuntimeException ignored) {
                // Try another factory or constructor.
            }
        }
        for (Constructor<?> constructor : tooltipType.getDeclaredConstructors()) {
            Class<?>[] parameterTypes = constructor.getParameterTypes();
            if (parameterTypes.length == 1
                    && parameterTypes[0].isAssignableFrom(message.getClass())) {
                try {
                    constructor.setAccessible(true);
                    return constructor.newInstance(message);
                } catch (ReflectiveOperationException | RuntimeException ignored) {
                    return null;
                }
            }
        }
        return null;
    }

    private static Object callback(Class<?> callbackType, Consumer<Button> onPress, Component message) {
        return Proxy.newProxyInstance(
                callbackType.getClassLoader(),
                new Class<?>[]{callbackType},
                (proxy, method, args) -> {
                    if (method.getDeclaringClass() == Object.class) {
                        return switch (method.getName()) {
                            case "toString" -> "e4steam Steam button callback";
                            case "hashCode" -> System.identityHashCode(proxy);
                            case "equals" -> proxy == (args == null ? null : args[0]);
                            default -> null;
                        };
                    }
                    if (args != null) {
                        for (Object argument : args) {
                            if (argument instanceof Button button) {
                                onPress.accept(button);
                                break;
                            }
                        }
                    }
                    return defaultValue(method.getReturnType(), message);
                }
        );
    }

    private static Object defaultValue(Class<?> type, Component message) {
        if (!type.isPrimitive()) {
            if (type.isAssignableFrom(message.getClass()) || type.isAssignableFrom(Component.class)) {
                return message;
            }
            if (type.isInterface()) {
                return Proxy.newProxyInstance(
                        type.getClassLoader(),
                        new Class<?>[]{type},
                        (proxy, method, args) -> defaultValue(method.getReturnType(), message)
                );
            }
            return null;
        }
        if (type == boolean.class) {
            return false;
        }
        if (type == char.class) {
            return '\0';
        }
        if (type == byte.class) {
            return (byte) 0;
        }
        if (type == short.class) {
            return (short) 0;
        }
        if (type == int.class) {
            return 0;
        }
        if (type == long.class) {
            return 0L;
        }
        if (type == float.class) {
            return 0F;
        }
        if (type == double.class) {
            return 0D;
        }
        return null;
    }
}

File: common/src/main/java/link/e4steam/MinecraftUiCompat.kt
package link.e4steam

import net.minecraft.client.Minecraft
import net.minecraft.client.gui.components.Button
import net.minecraft.client.gui.screens.ConnectScreen
import net.minecraft.client.gui.screens.Screen
import net.minecraft.client.multiplayer.ServerData
import net.minecraft.client.multiplayer.resolver.ServerAddress
import net.minecraft.network.chat.Component
import java.lang.reflect.Constructor
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.lang.reflect.Modifier
import java.lang.reflect.Proxy
import java.util.Comparator
import java.util.Optional
import java.util.function.Consumer

/**
 * Small reflection boundary for GUI APIs that changed after Minecraft 1.17.
 *
 * <p>The main JAR is compiled against 1.20.2, just like upstream e4steam's wide
 * compatibility build. Keeping Button.Builder and Tooltip out of bytecode
 * descriptors lets the same classes load on older Minecraft releases.</p>
 */
object MinecraftUiCompat {
    private val MESSAGE_SCREEN_CLASS_NAMES = arrayOf(
        // Mojang mappings through 1.20.6.
        "net.minecraft.client.gui.screens.GenericDirtMessageScreen",
        // Mojang mappings from 1.21 onward.
        "net.minecraft.client.gui.screens.GenericMessageScreen",
        // Stable Fabric intermediary name used across both eras.
        "net.minecraft.class_424"
    )

    private val DISCONNECT_METHOD_NAMES = arrayOf(
        "disconnect",
        "clearLevel",
        "method_18099",
        "method_18096",
        "m_91387_",
        "m_91398_"
    )

    @JvmStatic
    fun button(
        message: Component,
        onPress: Consumer<Button>,
        x: Int,
        y: Int,
        width: Int,
        height: Int
    ): Button {
        val fromBuilder = tryBuilder(message, onPress, x, y, width, height)
        if (fromBuilder != null) {
            return fromBuilder
        }

        val fromConstructor = tryConstructor(message, onPress, x, y, width, height)
        if (fromConstructor != null) {
            return fromConstructor
        }

        throw IllegalStateException("No compatible Minecraft Button factory was found")
    }

    /**
     * Adds a modern hover tooltip where that API exists. Minecraft 1.17-1.19.2
     * simply keeps the button without a tooltip.
     */
    @JvmStatic
    fun tooltip(widget: Any?, message: Component?) {
        if (widget == null || message == null) {
            return
        }

        for (setter in widget.javaClass.methods) {
            if (setter.parameterCount != 1 || Modifier.isStatic(setter.modifiers)) {
                continue
            }
            val tooltipType = setter.parameterTypes[0]
            val tooltip = createTooltip(tooltipType, message)
            if (tooltip == null) {
                continue
            }
            try {
                setter.invoke(widget, tooltip)
            } catch (_: ReflectiveOperationException) {
                // A tooltip is optional; the button itself remains usable.
                continue
            }
            return
        }
    }

    /**
     * Creates Minecraft's generic progress/message screen without linking the
     * caller to the Mojang class name that changed in 1.21.
     */
    @JvmStatic
    fun messageScreen(message: Component, fallback: Screen): Screen {
        for (className in MESSAGE_SCREEN_CLASS_NAMES) {
            try {
                val screenType = Class.forName(className, false, Screen::class.java.classLoader)
                if (!Screen::class.java.isAssignableFrom(screenType)) {
                    continue
                }
                for (constructor in screenType.declaredConstructors) {
                    val parameterTypes = constructor.parameterTypes
                    if (parameterTypes.size != 1
                        || !parameterTypes[0].isAssignableFrom(message.javaClass)) {
                        continue
                    }
                    constructor.isAccessible = true
                    return constructor.newInstance(message) as Screen
                }
            } catch (_: ClassNotFoundException) {
                // Try the next mapping/runtime name.
            } catch (_: ReflectiveOperationException) {
                // A progress screen is cosmetic. Keep joining via the previous
                // screen if this Minecraft version changes its constructor.
                break
            } catch (_: RuntimeException) {
                break
            }
        }
        return fallback
    }

    /**
     * Starts Minecraft's normal connection screen across its old four-argument
     * and newer five-/six-argument signatures.
     */
    @JvmStatic
    fun connect(
        parent: Screen,
        minecraft: Minecraft,
        address: ServerAddress,
        displayName: String,
        endpoint: String
    ) {
        val serverData = createServerData(displayName, endpoint)
        val methods = ConnectScreen::class.java.declaredMethods
        methods.sortWith(
            Comparator.comparingInt<Method> { method ->
                if (Modifier.isPublic(method.modifiers)) 0 else 1
            }
                .thenComparingInt { it.parameterCount }
        )
        for (method in methods) {
            if (!Modifier.isStatic(method.modifiers) || method.returnType != Void.TYPE) {
                continue
            }
            val types = method.parameterTypes
            if (!containsAssignable(types, Screen::class.java)
                || !containsAssignable(types, Minecraft::class.java)
                || !containsAssignable(types, ServerAddress::class.java)
                || !containsAssignable(types, ServerData::class.java)) {
                continue
            }

            val arguments = arrayOfNulls<Any?>(types.size)
            var valid = true
            for (i in types.indices) {
                val type = types[i]
                arguments[i] = when {
                    type.isInstance(parent) || type == Screen::class.java -> parent
                    type.isInstance(minecraft) || type == Minecraft::class.java -> minecraft
                    type.isInstance(address) || type == ServerAddress::class.java -> address
                    type.isInstance(serverData) || type == ServerData::class.java -> serverData
                    type == Boolean.TYPE -> false
                    type == Integer.TYPE -> 0
                    type == Long.TYPE -> 0L
                    type == Optional::class.java -> Optional.empty<Any>()
                    !type.isPrimitive -> null
                    else -> {
                        valid = false
                        break
                    }
                }
            }
            if (!valid) {
                continue
            }
            method.isAccessible = true
            method.invoke(null, *arguments)
            return
        }
        throw NoSuchMethodException("No compatible ConnectScreen entry point was found")
    }

    /** Disconnects the current world across the old clearLevel/new disconnect rename. */
    @JvmStatic
    fun disconnect(minecraft: Minecraft, nextScreen: Screen) {
        for (name in DISCONNECT_METHOD_NAMES) {
            for (method in Minecraft::class.java.methods) {
                if (!method.name.equals(name)
                    || Modifier.isStatic(method.modifiers)
                    || method.returnType != Void.TYPE
                    || method.parameterCount != 1
                    || !method.parameterTypes[0].isAssignableFrom(nextScreen.javaClass)) {
                    continue
                }
                method.invoke(minecraft, nextScreen)
                return
            }
        }
        throw NoSuchMethodException("No compatible Minecraft disconnect method was found")
    }

    private fun tryBuilder(
        message: Component,
        onPress: Consumer<Button>,
        x: Int,
        y: Int,
        width: Int,
        height: Int
    ): Button? {
        for (factory in Button::class.java.declaredMethods) {
            if (!Modifier.isStatic(factory.modifiers) || factory.parameterCount != 2) {
                continue
            }
            val parameters = factory.parameterTypes
            if (!parameters[0].isAssignableFrom(message.javaClass)
                && !parameters[0].isAssignableFrom(Component::class.java)) {
                continue
            }
            if (!parameters[1].isInterface) {
                continue
            }

            val callback = callback(parameters[1], onPress, message)
            try {
                factory.isAccessible = true
                val builder = factory.invoke(null, message, callback)
                if (builder == null || builder is Button) {
                    continue
                }

                val bounds = builder.javaClass.methods
                    .filter { !Modifier.isStatic(it.modifiers) }
                    .filter { it.parameterCount == 4 }
                    .filter { it.parameterTypes.all { type -> type == Integer.TYPE } }
                    .filter { method ->
                        method.returnType.isAssignableFrom(builder.javaClass)
                            || builder.javaClass.isAssignableFrom(method.returnType)
                    }
                    .firstOrNull()
                if (bounds == null) {
                    continue
                }
                bounds.invoke(builder, x, y, width, height)

                val build = builder.javaClass.methods
                    .filter { !Modifier.isStatic(it.modifiers) }
                    .filter { it.parameterCount == 0 }
                    .filter { Button::class.java.isAssignableFrom(it.returnType) }
                    .firstOrNull()
                if (build != null) {
                    return build.invoke(builder) as Button
                }
            } catch (_: ReflectiveOperationException) {
                // Fall through to the pre-builder constructor path.
            } catch (_: RuntimeException) {
                // Fall through to the pre-builder constructor path.
            }
        }
        return null
    }

    private fun createServerData(displayName: String, endpoint: String): Any {
        val constructors = ServerData::class.java.declaredConstructors
        constructors.sortWith(Comparator.comparingInt<Constructor<*>> { it.parameterCount })
        var lastFailure: ReflectiveOperationException? = null
        for (constructor in constructors) {
            val types = constructor.parameterTypes
            if (types.size < 2 || types[0] != String::class.java || types[1] != String::class.java) {
                continue
            }
            val arguments = arrayOfNulls<Any?>(types.size)
            arguments[0] = displayName
            arguments[1] = endpoint
            for (i in 2 until types.size) {
                val type = types[i]
                arguments[i] = when {
                    type == Boolean.TYPE -> false
                    type == Integer.TYPE -> 0
                    type == Long.TYPE -> 0L
                    type.isEnum -> {
                        val constants = type.enumConstants!!
                        var selected: Any? = null
                        for (constant in constants) {
                            if ((constant as Enum<*>).name == "OTHER") {
                                selected = constant
                                break
                            }
                        }
                        selected ?: (if (constants.isEmpty()) null else constants[0])
                    }
                    type == Optional::class.java -> Optional.empty<Any>()
                    else -> null
                }
            }
            try {
                constructor.isAccessible = true
                return constructor.newInstance(*arguments)
            } catch (failure: ReflectiveOperationException) {
                lastFailure = failure
            }
        }
        if (lastFailure != null) {
            throw lastFailure!!
        }
        throw NoSuchMethodException("No compatible ServerData constructor was found")
    }

    private fun containsAssignable(haystack: Array<Class<*>>, needle: Class<*>): Boolean {
        for (type in haystack) {
            if (type != Any::class.java
                && (type.isAssignableFrom(needle) || needle.isAssignableFrom(type))) {
                return true
            }
        }
        return false
    }

    private fun tryConstructor(
        message: Component,
        onPress: Consumer<Button>,
        x: Int,
        y: Int,
        width: Int,
        height: Int
    ): Button? {
        val constructors = Button::class.java.declaredConstructors
        constructors.sortWith(Comparator.comparingInt<Constructor<*>> { it.parameterCount })
        for (constructor in constructors) {
            val parameterTypes = constructor.parameterTypes
            if (parameterTypes.size < 6
                || parameterTypes[0] != Integer.TYPE
                || parameterTypes[1] != Integer.TYPE
                || parameterTypes[2] != Integer.TYPE
                || parameterTypes[3] != Integer.TYPE
                || !Component::class.java.isAssignableFrom(parameterTypes[4])
                || !parameterTypes[5].isInterface) {
                continue
            }

            val arguments = arrayOfNulls<Any?>(parameterTypes.size)
            arguments[0] = x
            arguments[1] = y
            arguments[2] = width
            arguments[3] = height
            arguments[4] = message
            arguments[5] = callback(parameterTypes[5], onPress, message)
            for (i in 6 until parameterTypes.size) {
                arguments[i] = defaultValue(parameterTypes[i], message)
            }

            try {
                constructor.isAccessible = true
                return constructor.newInstance(*arguments) as Button
            } catch (_: ReflectiveOperationException) {
                // Try the next constructor shape.
            } catch (_: RuntimeException) {
                // Try the next constructor shape.
            }
        }
        return null
    }

    private fun createTooltip(tooltipType: Class<*>, message: Component): Any? {
        for (method in tooltipType.declaredMethods) {
            if (!Modifier.isStatic(method.modifiers)
                || method.parameterCount != 1
                || !tooltipType.isAssignableFrom(method.returnType)
                || !method.parameterTypes[0].isAssignableFrom(message.javaClass)) {
                continue
            }
            try {
                method.isAccessible = true
                return method.invoke(null, message)
            } catch (_: ReflectiveOperationException) {
                // Try another factory or constructor.
            } catch (_: RuntimeException) {
                // Try another factory or constructor.
            }
        }
        for (constructor in tooltipType.declaredConstructors) {
            val parameterTypes = constructor.parameterTypes
            if (parameterTypes.size == 1
                && parameterTypes[0].isAssignableFrom(message.javaClass)) {
                try {
                    constructor.isAccessible = true
                    return constructor.newInstance(message)
                } catch (_: ReflectiveOperationException) {
                    return null
                } catch (_: RuntimeException) {
                    return null
                }
            }
        }
        return null
    }

    private fun callback(callbackType: Class<*>, onPress: Consumer<Button>, message: Component): Any {
        return Proxy.newProxyInstance(
            callbackType.classLoader,
            arrayOf<Class<*>>(callbackType),
            InvocationHandler { proxy, method, args ->
                if (method.declaringClass == Any::class.java) {
                    when (method.name) {
                        "toString" -> "e4steam Steam button callback"
                        "hashCode" -> System.identityHashCode(proxy)
                        "equals" -> proxy == args?.getOrNull(0)
                        else -> null
                    }
                } else {
                    if (args != null) {
                        for (argument in args) {
                            if (argument is Button) {
                                onPress.accept(argument)
                                break
                            }
                        }
                    }
                    defaultValue(method.returnType, message)
                }
            }
        )
    }

    private fun defaultValue(type: Class<*>, message: Component): Any? {
        if (!type.isPrimitive) {
            if (type.isAssignableFrom(message.javaClass) || type.isAssignableFrom(Component::class.java)) {
                return message
            }
            if (type.isInterface) {
                return Proxy.newProxyInstance(
                    type.classLoader,
                    arrayOf<Class<*>>(type),
                    InvocationHandler { _, method, _ -> defaultValue(method.returnType, message) }
                )
            }
            return null
        }
        return when (type) {
            Boolean.TYPE -> false
            Character.TYPE -> '\u0000'
            Byte.TYPE -> 0.toByte()
            Short.TYPE -> 0.toShort()
            Integer.TYPE -> 0
            Long.TYPE -> 0L
            Float.TYPE -> 0F
            Double.TYPE -> 0D
            else -> null
        }
    }
}
